File: leetcode124/leetcode124.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        self.max_val = float("-inf")
        def dfs(root):
            if not root:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)
            self.max_val = max(self.max_val, root.val + left + right)
            outSum = root.val + max(0, max(left, right))
  
            return max(outSum, 0)
        dfs(root)
        return self.max_val

File: leetcode124/leetcode124.cpp
#include<bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int ans = INT_MIN;
    int dfs(TreeNode* root){
        if (root == nullptr) return 0;
        int left = dfs(root->left);
        int right = dfs(root->right);

        int inSum = root->val + left + right;
        ans = max(ans, inSum);
        int outSum = root->val + max(0, max(left, right));
        return max(outSum, 0);
    }
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return ans;
    }
};